eva/optimizer/rules/rules.py

```diff
@@ -317,6 +319,99 @@ def apply(self, before: LogicalFilter, context: OptimizerContext):
         return new_join_node
 
 
+class XformLateralJoinToLinearFlow(Rule):
+    """If the inner node of a lateral join is a function-valued expression, we
+    eliminate the join node and make the inner node the parent of the outer node. This
+    produces a linear #data flow path. Because this scenario is common in our system,
+    we chose to explicitly convert it to a linear flow, which simplifies the
+    implementation of other optimizations such as UDF reuse and parallelized plans by
+    removing the join."""
+
+    def __init__(self):
+        pattern = Pattern(OperatorType.LOGICALJOIN)
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        pattern.append_child(Pattern(OperatorType.LOGICALFUNCTIONSCAN))
+        super().__init__(RuleType.XFORM_LATERAL_JOIN_TO_LINEAR_FLOW, pattern)
+
+    def promise(self):
+        return Promise.XFORM_LATERAL_JOIN_TO_LINEAR_FLOW
+
+    def check(self, before: LogicalJoin, context: OptimizerContext):
+        if before.join_type == JoinType.LATERAL_JOIN:
+            if before.join_predicate is None and not before.join_project:
+                return True
+        return False
+
+    def apply(self, before: LogicalJoin, context: OptimizerContext):
+        #     LogicalJoin(Lateral)              LogicalApplyAndMerge
+        #     /           \                 ->       |
+        #    A        LogicalFunctionScan            A
+
+        A: Dummy = before.children[0]
+        logical_func_scan: LogicalFunctionScan = before.children[1]
+        logical_apply_merge = LogicalApplyAndMerge(
+            logical_func_scan.func_expr,
+            logical_func_scan.alias,
+            logical_func_scan.do_unnest,
+        )
+        logical_apply_merge.append_child(A)
+        return logical_apply_merge
+
+
+class PushDownFilterThroughApplyAndMerge(Rule):
+    """If it is feasible to partially or fully push the predicate contained within the
+    logical filter through the ApplyAndMerge operator, we should do so. This is often
+    beneficial, for instance, in order to prevent decoding additional frames beyond
+    those that satisfy the predicate.
+    Eg:
+
+    Filter(id < 10 and func.label = 'car')           Filter(func.label = 'car')
+            |                                                   |
+        ApplyAndMerge(func)                  ->          ApplyAndMerge(func)
+            |                                                   |
+            A                                            Filter(id < 10)
+                                                                |
```

Comment:
Does push `label='car'` into the `ApplyAndMerge` help? So the join can work on few rows and the output batch is smaller. We can defer this optimization to later PRs. Just want to confirm whether we should have a rule for that.

Reply:
It might not make much of a difference as we aren't doing join anymore. It is more like a merge. However, we can profile and add it later.